startup/11-optics.py

```diff
@@ -0,0 +1,173 @@
+print(f'Loading {__file__}...')
+
+import numpy as np
+from ophyd import EpicsMotor, EpicsSignal, EpicsSignalRO
+from ophyd import Device
+from ophyd import Component as Cpt
+from ophyd import PVPositionerPC
+from nslsii.devices import TwoButtonShutter
+
+
+### Setup photon shutters
+shut_fe = TwoButtonShutter('XF:05ID-PPS{Sh:WB}', name='shut_fe')
+shut_a = TwoButtonShutter('XF:05IDA-PPS:1{PSh:2}', name='shut_a')
+shut_b = TwoButtonShutter('XF:05IDB-PPS:1{PSh:4}', name='shut_b')
+
+
+### Setup white/pink beam slits
+class SRXSlitsWB(Device):
+    # Real synthetic axes
+    h_cen = Cpt(EpicsMotor, 'XCtr}Mtr')
+    h_gap = Cpt(EpicsMotor, 'XGap}Mtr')
+    v_cen = Cpt(EpicsMotor, 'YCtr}Mtr')
+    v_gap = Cpt(EpicsMotor, 'YGap}Mtr')
+
+    # Real motors
+    top = Cpt(EpicsMotor, 'T}Mtr')
+    bot = Cpt(EpicsMotor, 'B}Mtr')
+    inb = Cpt(EpicsMotor, 'I}Mtr')
+    out = Cpt(EpicsMotor, 'O}Mtr')
+
+class SRXSlitsPB(Device):
+    # Real synthetic axes
+    h_cen = Cpt(EpicsMotor, 'XCtr}Mtr')
+    h_gap = Cpt(EpicsMotor, 'XGap}Mtr')
+
+    # Real motors
+    inb = Cpt(EpicsMotor, 'I}Mtr')
+    out = Cpt(EpicsMotor, 'O}Mtr')
+
+slt_wb = SRXSlitsWB('XF:05IDA-OP:1{Slt:1-Ax:', name='slt_wb')
+slt_pb = SRXSlitsPB('XF:05IDA-OP:1{Slt:2-Ax:', name='slt_pb')
+
+# Is this used? Commented out before removing
+# class SRXSlits2(Device):
+#     inb = Cpt(EpicsMotor, 'I}Mtr')
+#     out = Cpt(EpicsMotor, 'O}Mtr')
+
+
+### Setup HFM Mirror
+class SRXHFM(Device):
+    x = Cpt(EpicsMotor, 'X}Mtr')
+    y = Cpt(EpicsMotor, 'Y}Mtr')
+    pitch = Cpt(EpicsMotor, 'P}Mtr')
+    bend = Cpt(EpicsMotor, 'Bend}Mtr')
+
+hfm = SRXHFM('XF:05IDA-OP:1{Mir:1-Ax:', name='hfm')
+
+
+### Setup HDCM
+class HDCMPIEZOROLL(PVPositionerPC):
+    setpoint = Cpt(EpicsSignal, '')
+    readback = Cpt(EpicsSignalRO, '')
+    pid_enabled = EpicsSignal('XF:05IDD-CT{FbPid:01}PID:on')
+    pid_I = EpicsSignal('XF:05IDD-CT{FbPid:01}PID.I')
+
+    def reset_pid(self):
+        yield from bps.mov(self.pid_I, 0.0)
+
+class HDCMPIEZOPITCH(PVPositionerPC):
+    setpoint = Cpt(EpicsSignal, '')
+    readback = Cpt(EpicsSignalRO, '')
+    pid_enabled = EpicsSignal('XF:05IDD-CT{FbPid:02}PID:on')
+    pid_I = EpicsSignal('XF:05IDD-CT{FbPid:02}PID.I')
+
+    def reset_pid(self):
+        yield from bps.mov(self.pid_I, 0.0)
+
+class SRXDCM(Device):
+    bragg = Cpt(EpicsMotor, 'P}Mtr')
+    c1_roll = Cpt(EpicsMotor, 'R1}Mtr')
+    c1_fine = HDCMPIEZOROLL('XF:05ID-BI{EM:BPM1}DAC0', name='c1_roll')
+    c2_x = Cpt(EpicsMotor, 'X2}Mtr')
+    c2_pitch = Cpt(EpicsMotor, 'P2}Mtr')
+    c2_fine = HDCMPIEZOPITCH('XF:05ID-BI{EM:BPM1}DAC1', name='c2_fine')
+    c2_pitch_kill = Cpt(EpicsSignal, 'P2}Cmd:Kill-Cmd')
+    x = Cpt(EpicsMotor, 'X}Mtr')
+    y = Cpt(EpicsMotor, 'Y}Mtr')
+
+    temp_pitch = Cpt(EpicsSignalRO, 'P}T-I')
+
+dcm = SRXDCM('XF:05IDA-OP:1{Mono:HDCM-Ax:' , name='dcm')
```

Comment:
As we discussed, this may be used as a component for the energy pseudo-positioner.

Reply:
Tried to do this but kept getting errors in bluesky. I basically tried inside the energy class to do "bragg = dcm.bragg", but I kept getting errors in the inverse function of energy. 

AttributeError: 'EnergyRealPos' object has no attribute 'bragg'
Subscription readback callback exception (EpicsMotor(prefix='SR:C5-ID:G1{IVU21:1-Ax:Gap}-Mtr', name='energy_u_gap_gap', parent='energy_u_gap', settle_time=0.0, timeout=None, read_attrs=['user_readback', 'user_setpoint'], configuration_attrs=['user_offset', 'user_offset_dir', 'velocity', 'acceleration', 'motor_egu']))